sql/cagg_api.sql

```diff
@@ -0,0 +1,162 @@
+        USING ERRCODE = 'wrong_object_type';
+    END IF;
+
+    RETURN info;
+END
+$body$
+LANGUAGE plpgsql
+SET search_path TO pg_catalog, pg_temp;
+
+-- Add new invalidations to the materialization invalidation log.
+--
+-- This will add the range to the materialization invalidations for
+-- the continuous aggregate. The range will automatically be "aligned"
+-- to the bucket width to ensure that it covers all buckets that it
+-- touches.
+CREATE OR REPLACE PROCEDURE _timescaledb_functions.add_materialization_invalidations(
+    continuous_aggregate regclass,
+    invalidation tsrange
+) AS
+$body$
+DECLARE
+    info RECORD := _timescaledb_functions.get_materialization_info(continuous_aggregate);
+    aligned TSRANGE := _timescaledb_functions.align_to_bucket(info.bucket_width, invalidation);
+BEGIN
+    INSERT INTO _timescaledb_catalog.continuous_aggs_materialization_invalidation_log
+    VALUES (info.materialization_id,
+            _timescaledb_functions.to_unix_microseconds(lower(aligned)),
+            _timescaledb_functions.to_unix_microseconds(upper(aligned)));
+END
+$body$
+LANGUAGE plpgsql
+SET search_path TO pg_catalog, pg_temp;
+
+CREATE OR REPLACE PROCEDURE _timescaledb_functions.add_materialization_invalidations(
+    continuous_aggregate REGCLASS,
+    invalidation TSTZRANGE
+) AS
+$body$
+DECLARE
+    info RECORD := _timescaledb_functions.get_materialization_info(continuous_aggregate);
+    aligned TSTZRANGE := _timescaledb_functions.align_to_bucket(info.bucket_width, invalidation);
+BEGIN
+    INSERT INTO _timescaledb_catalog.continuous_aggs_materialization_invalidation_log
+    VALUES (info.materialization_id,
+            _timescaledb_functions.to_unix_microseconds(lower(aligned)),
+            _timescaledb_functions.to_unix_microseconds(upper(aligned)));
+END
+$body$
+LANGUAGE plpgsql
+SET search_path TO pg_catalog, pg_temp;
+
+-- Get raw ranges from the materialization invalidation log
+--
+-- This is a cleaned-up version of the timestamps, still in Unix
+-- microseconds, with nulls for '-infinity' and '+infinity' and
+-- invalid entries removed.
+CREATE OR REPLACE FUNCTION _timescaledb_functions.get_raw_materialization_ranges(typ regtype)
+RETURNS TABLE (materialization_id integer,
+               lowest_modified_value bigint,
+               greatest_modified_value bigint)
+AS $$
+   SELECT materialization_id,
+          NULLIF(lowest_modified_value, _timescaledb_functions.get_internal_time_min(typ)),
+          NULLIF(greatest_modified_value, _timescaledb_functions.get_internal_time_max(typ))
+     FROM _timescaledb_catalog.continuous_aggs_materialization_invalidation_log
+    WHERE lowest_modified_value
+          BETWEEN _timescaledb_functions.get_internal_time_min(typ)
+              AND _timescaledb_functions.get_internal_time_max(typ)
+      AND greatest_modified_value
+          BETWEEN _timescaledb_functions.get_internal_time_min(typ)
+              AND _timescaledb_functions.get_internal_time_max(typ)
```

Comment:
```suggestion
   WITH min_max_values AS (
      SELECT _timescaledb_functions.get_internal_time_min(typ) AS min,
      _timescaledb_functions.get_internal_time_max(typ) AS max
   )
   SELECT materialization_id,
          NULLIF(lowest_modified_value, min_max_values.min),
          NULLIF(greatest_modified_value, min_max_values.max)
     FROM _timescaledb_catalog.continuous_aggs_materialization_invalidation_log, min_max_values
    WHERE lowest_modified_value
          BETWEEN min_max_values.min
              AND min_max_values.max
      AND greatest_modified_value
          BETWEEN min_max_values.min
              AND min_max_values.max
```
Doing this way will avoid so many function calls.